### ISSUE TITLE
Two Empty Buttons Causing Accessibility Issue

### ISSUE THREAD

**Issue by teslajanny** [FrDH/jQuery.mmenu#817]
Hi there. There are two empty buttons; one with a class of mm-tabstart and the other with a class of mm-tabend. These buttons do not have any text inside them which causes the WAVE toolbar to flag them as an accessibility issue.

`<button class="mm-tabstart" tabindex="0" type="button" aria-hidden="true"></button>`

`<button class="mm-tabend" tabindex="0" type="button" aria-hidden="true"></button>`

There should be text inside these buttons to solve the accessibility issue and to make WAVE happy.

Thanks.

**Comment by FrDH** [comment]
These buttons are added by the keyboardNavigation add-on to restrict the focussable elements using the tab-button. I guess I could add tekst to the buttons, but I don't think it would add any accessibility, it would only be to make WAVE happy.

**Comment by teslajanny** [comment]
Thank you for your response!

What are the buttons actually used for and who uses them (like what kind of user)? Testing it now, the menu is keyboard accessible (we are using the D8 responsive_menu module which requires your mmenu library). 

I'm sure WAVE isn't the only accessibility tool that will flag the empty buttons. We are just trying to meet accessibility requirements which, unfortunately, flag the buttons even though they aren't read by screen readers.

**Comment by FrDH** [comment]
The buttons are used to restrict the focussable elements to those that are actually accessible for the user. Without the buttons, you'd be able to set focus on an element outside of the menu.

**Comment by geckoseo** [comment]
Hi, this is technically an accessibility issue for screen readers, and it can cause issues for organizations who are trying to meet regulatory requirements to ensure that their websites are accessible (and it has an SEO impact also).

Here is an example that helps meet the accessibility requirements.

It needs to be added into the html markup of the visually hidden object since obviously javascript injection isn't seen by screen readers until it is too late.

`<button class="visuallyhidden mm-tabstart" tabindex="0" type="button" aria-hidden="true">SCROLL LOCK</button>

<button class="visuallyhidden mm-tabend" tabindex="0" type="button" aria-hidden="true">SCROLL RELEASE</button>`

`.visuallyhidden {
    border: 0;
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
}`

**Comment by FrDH** [comment]
@geckoseo I don't think we're talking about the same buttons.

**Comment by geckoseo** [comment]
@FrDH yes.. we are talking about the same buttons.. but that doesn't mean my example will solve the accessibility issue.

The idea was pointing out the <aria-hidden="true"> attribute, along with adding text to the buttons that can be read by screenreaders, but hidden by adding a class that clips and collapses the text so that it is invisible to eyes.

**Comment by FrDH** [comment]
The thing is, screenreaders do not need to read anything on those buttons. The buttons are only there to prevent tabbing outside the menu and current opened panel.

**Comment by geckoseo** [comment]
The screenreaders don't "need" to read them, correct, however screenreaders do read them, and this can cause confusion for people who are disabled (blind people for example) because the technology they are using to read the content for them doesn't always know that hidden buttons are fulfilling a function that is not related to the visible content.

For websites that need to meet regulatory accessibility requirements, if the invisible buttons don't pass the strict requirements of accessibility, then the developers of a website need to ensure they meet the legal requirements or they can be punished (i.e, a fine, or in severe breaches, a site must be closed until the requirements are met).

It's not your job to make sure the tool is perfect, although of course people always appreciate tools that require little or no modification to use for their own unique requirements.

Really all I was offering was a potential method for ensuring that the hidden buttons do not cause screen readers to try and make sense of the buttons and/or read content back to the blind person that makes no sense to them because they can't actually see the screen and are relying entirely on the screen reader to read the content out loud.

I have not tested the possible solution, because I don't need to use it. It's just information for people to consider using if they would like to use this menu, and if they run into an accessibility requirement, which is what I understood the original issue to be about.

**Comment by FrDH** [comment]
Is there a way to let screen readers ignore the buttons? That would be best I think.

What did yoe mean with
> It needs to be added into the html markup of the visually hidden object since obviously javascript injection isn't seen by screen readers until it is too late.

The whole menu -including the two buttons- is created with JS...
